In [1]:
#### Importing the libraries ####
import pandas as pd
import numpy as np
import random

#### All rights reserved. DataPoint Armenia 2021.

![title](amberd.png)

# Data Cleaning

#### 1. Read the dataset.

In [2]:
#### reading the dataset ####

diabetes = pd.read_csv("data/dirty_diabetes.csv")

#### head() shows the first 5 rows
diabetes.head()

Unnamed: 0  Unnamed: 0.1  Age Gender Polyuria Polydipsia  \
0           0             0   40   Male       No        Yes   
1           1             1   58   Male       No         No   
2           2             2   41   Male      Yes         No   
3           3             3   45   Male       No         No   
4           4             4   60   Male      Yes        Yes   

  sudden weight loss weakness Polyphagia Genital thrush  ... Irritability  \
0                 No      Yes         No             No  ...           No   
1                 No      Yes         No             No  ...           No   
2                 No      Yes        Yes             No  ...           No   
3                Yes      Yes        Yes            Yes  ...           No   
4                Yes      Yes        Yes             No  ...          Yes   

  delayed healing partial paresis muscle stiffness Alopecia Obesity     class  \
0             Yes              No              Yes      Yes     Yes  Positive   
1              No             Yes               No      Yes      No  Positive   
2             Yes              No              Yes      Yes      No  Positive   
3             Yes              No               No       No      No  Positive   
4             Yes             Yes              Yes      Yes     Yes  Positive   

     names     id Times Pregnant  
0     Liam  18989            0.0  
1     Noah  15244            NaN  
2   Oliver  12575            0.0  
3  William  14613            0.0  
4   Elijah  12680           99.0  

[5 rows x 22 columns]

## Exploration

#### 2. Check the number of observations.

In [71]:
#### check the number of observations in the dataset 
len(diabetes)

540

#### 3. Check the columns.


In [72]:
### check the columns
diabetes.columns

Index(['Unnamed: 0', 'Unnamed: 0.1', 'Age', 'Gender', 'Polyuria', 'Polydipsia',
       'sudden weight loss', 'weakness', 'Polyphagia', 'Genital thrush',
       'visual blurring', 'Itching', 'Irritability', 'delayed healing',
       'partial paresis', 'muscle stiffness', 'Alopecia', 'Obesity', 'class',
       'names', 'id', 'Times Pregnant'],
      dtype='object')

#### 4. Use decribe() for getting some sense of the data.

In [3]:
####  YOUR CODE

### outputs only the numeric columns

#### 5. Look for irrelavant columns. What is Unnamed: 0?

In [74]:
#### Look for irrelavant columns or values

#### YOUR CODE
diabetes['Unnamed: 0']

0        0
1        1
2        2
3        3
4        4
      ... 
535    535
536    536
537    537
538    538
539    539
Name: Unnamed: 0, Length: 540, dtype: int64

#### 6. Look for irrelavant columns. What is Unnamed: 1?

In [75]:
### what is Unnamed: 0.1
diabetes['Unnamed: 0.1']

0        0
1        1
2        2
3        3
4        4
      ... 
535    167
536    400
537    163
538    427
539    212
Name: Unnamed: 0.1, Length: 540, dtype: int64

#### 7. Drop the irrelevant columns.

In [4]:
### drop the irrelevan columns
### YOUR CODE HERE



#### 8. Who had been pregnant for 99 times?

In [80]:
#### Who had been pregnant for 99 times?
diabetes[diabetes['Times Pregnant'] == 99]

Age Gender Polyuria Polydipsia sudden weight loss weakness Polyphagia  \
4   60   Male      Yes        Yes                Yes      Yes        Yes   

  Genital thrush visual blurring Itching Irritability delayed healing  \
4             No             Yes     Yes          Yes             Yes   

  partial paresis muscle stiffness Alopecia Obesity     class   names     id  \
4             Yes              Yes      Yes     Yes  Positive  Elijah  12680   

   Times Pregnant  
4            99.0

#### 9. Drop the row corresponding to 99 pregnancies.

In [5]:
#### YOUR CODE


#### 10. Why is the count in Times Pregnant 451 while the others have 540? Check for NaN values using isnull() .

In [7]:
#### YOUR CODE


#### 11. Do we drop the missing names? Will your analysis suffer from this missingness?

#### 12. Let's explore the missingness in Times Pregnant column. Should we drop all the NAs in Times Pregnant column? 

Can we predict the number of Times Pregnant from the other columns? 

HINT: use the Gender column to predict some of the Times Pregnan values

In [104]:
diabetes[['Gender', 'Times Pregnant']]

Gender  Times Pregnant
0      Male             0.0
1      Male             NaN
2      Male             0.0
3      Male             0.0
5      Male             0.0
..      ...             ...
535    Male             0.0
536    Male             0.0
537  Female             0.0
538    Male             0.0
539    Male             0.0

[539 rows x 2 columns]

#### 13. Impute 0s instead of NaNs for Male in the dataset and check the number of NaN values.

In [105]:
### impute 0s instead of NaNs for Male in the dataset
diabetes.loc[diabetes['Gender'] == 'Male', 'Times Pregnant'] = 0

In [8]:
### check!

### YOUR CODE


#### 14. Drop all the NaNs for Female.

In [109]:
### drop all the NaNs for Female
diabetes = diabetes.dropna(subset = ['Times Pregnant'])

In [110]:
### YAY!!!
diabetes['Times Pregnant'].isnull().sum()

0

#### 15. check for DUPLICATES!

HINT: check using the unique idetifier. Use duplicated() function with specifying the *subset*

In [9]:
#### YOUR CODE


In [115]:
### count the number of duplicates
diabetes.duplicated(subset = 'id').sum()

17

#### 16. Drop the DUPLICATES.

HINT: Use drop_duplicates() function with specifying the *subset*

In [10]:
### YOUR CODE


In [118]:
### CHECK
diabetes['Times Pregnant'].isnull().sum()
### YAY!

0

#### 17. Lastly, convert the types!

In [121]:
yes_no_dict = {'Yes': 1,
              'No': 0}
diabetes = diabetes.replace(yes_no_dict)
diabetes.head()


Age Gender  Polyuria  Polydipsia  sudden weight loss  weakness  Polyphagia  \
0   40   Male         0           1                   0         1           0   
1   58   Male         0           0                   0         1           0   
2   41   Male         1           0                   0         1           1   
3   45   Male         0           0                   1         1           1   
5   55   Male         1           1                   0         1           1   

   Genital thrush  visual blurring  Itching  Irritability  delayed healing  \
0               0                0        1             0                1   
1               0                1        0             0                0   
2               0                0        1             0                1   
3               1                0        1             0                1   
5               0                1        1             0                1   

   partial paresis  muscle stiffness  Alopecia  Obesity     class    names  \
0                0                 1         1        1  Positive     Liam   
1                1                 0         1        0  Positive     Noah   
2                0                 1         1        0  Positive   Oliver   
3                0                 0         0        0  Positive  William   
5                0                 1         1        1  Positive    James   

      id  Times Pregnant  
0  18989             0.0  
1  15244             0.0  
2  12575             0.0  
3  14613             0.0  
5  10125             0.0

In [123]:
### you can also use pd.get_dummies
pd.get_dummies(diabetes['Gender'], drop_first = True)


Male
0       1
1       1
2       1
3       1
5       1
..    ...
514     0
516     0
517     0
518     0
519     1

[481 rows x 1 columns]

In [124]:
### add it as a column
diabetes['Gender'] = pd.get_dummies(diabetes['Gender'], drop_first = True)
diabetes

Age  Gender  Polyuria  Polydipsia  sudden weight loss  weakness  \
0     40       1         0           1                   0         1   
1     58       1         0           0                   0         1   
2     41       1         1           0                   0         1   
3     45       1         0           0                   1         1   
5     55       1         1           1                   0         1   
..   ...     ...       ...         ...                 ...       ...   
514   54       0         1           1                   1         1   
516   48       0         1           1                   1         1   
517   58       0         1           1                   1         1   
518   32       0         0           0                   0         1   
519   42       1         0           0                   0         0   

     Polyphagia  Genital thrush  visual blurring  Itching  Irritability  \
0             0               0                0        1             0   
1             0               0                1        0             0   
2             1               0                0        1             0   
3             1               1                0        1             0   
5             1               0                1        1             0   
..          ...             ...              ...      ...           ...   
514           1               0                0        0             0   
516           1               0                0        1             1   
517           1               0                1        0             0   
518           0               0                1        1             0   
519           0               0                0        0             0   

     delayed healing  partial paresis  muscle stiffness  Alopecia  Obesity  \
0                  1                0                 1         1        1   
1                  0                1                 0         1        0   
2                  1                0                 1         1        0   
3                  1                0                 0         0        0   
5                  1                0                 1         1        1   
..               ...              ...               ...       ...      ...   
514                0                1                 0         0        0   
516                1                1                 0         0        0   
517                0                1                 1         0        1   
518                1                0                 0         1        0   
519                0                0                 0         0        0   

        class    names     id  Times Pregnant  
0    Positive     Liam  18989             0.0  
1    Positive     Noah  15244             0.0  
2    Positive   Oliver  12575             0.0  
3    Positive  William  14613             0.0  
5    Positive    James  10125             0.0  
..        ...      ...    ...             ...  
514  Positive   Jayson  14755             0.0  
516  Positive   Briggs  11239             4.0  
517  Positive  Armando  19096             3.0  
518  Negative  Nikolai  14091             3.0  
519  Negative  Cassius  13029             0.0  

[481 rows x 20 columns]

#### 18. Are we missing any other preprocessing step?
